src/timers.js

```diff
@@ -0,0 +1,150 @@
+const realSetTimeout = window.setTimeout;
+const realClearTimeout = window.clearTimeout;
+const realSetInterval = window.setInterval;
+const realClearInterval = window.clearInterval;
+
+// export real timers for testing
+export { realSetTimeout, realSetInterval };
+
+var runningTimeouts = [];
+var runningIntervals = [];
+
+/**
+ * Mark an timer as executed inside the test constraints
+ *
+ * @param {Number} timerId
+ */
+function removeTimeout(timerId) {
+  runningTimeouts = runningTimeouts.filter(({ id }) => id !== timerId);
+}
+
+/**
+ * Overridden setTimeout function that detects when it executes outside a testing constraint (`it`)
+ *
+ * @param {Function} fn
+ * @param {Array<*>} timerArgs
+ * @returns {Number}
+ */
+function localSetTimeout(fn, ...timerArgs) {
+  if (typeof fn === 'string') {
+    throw new Error('String function arguments for "setTimeout" cannot be executed.');
+  }
+
+  let timerId;
+  const timerFn = function(...args) {
+    removeTimeout(timerId);
+    return fn.apply(this, args);
+  };
+  timerId = realSetTimeout.apply(this, [timerFn, ...timerArgs]);
+  // must throw the error for PhantomJS to generate the stack trace
+  let err;
+  try {
+    throw new Error('Stray "setTimeout" call was executed outside the test constraints');
+  }
+  catch (e) {
+    err = e;
+  }
+  runningTimeouts.push({
+    id: timerId,
+    err
+  });
+  return timerId;
+}
+
+/**
+ * Overridden clearTimeout function that removes the check for a stray setTimeout
+ *
+ * @param {Number} timerId
+ * @returns {undefined}
+ */
+function localClearTimeout(timerId) {
+  removeTimeout(timerId);
+  return realClearTimeout.apply(this, arguments);
+}
+
+/**
+ * Mark an interval as cleared inside the test constraints
+ *
+ * @param {Number} timerId
+ */
+function removeInterval(timerId) {
+  runningIntervals = runningIntervals.filter(({ id }) => id !== timerId);
```

Comment:
`.filter(({ id }) => id !== timerId)` same is used on line 18, consider extracting to a function.